messaging/src/main/java/org/axonframework/eventhandling/pooled/WorkPackage.java

```diff
@@ -191,64 +198,73 @@ public boolean scheduleEvents(List<TrackedEventMessage<?>> events) {
-                      .map(TrackedEventMessage::trackingToken)
-                      .allMatch(token -> Objects.equals(expectedToken, token)),
+                eventEntries.stream()
+                            .map(entry -> TrackingToken.fromContext(entry).orElse(null))
+                            .allMatch(token -> Objects.equals(expectedToken, token)),
                 () -> "All tokens should match when scheduling multiple events in one go."
         );
     }
 
     /**
-     * Schedule a {@link TrackedEventMessage} for processing by this work package. Will immediately disregard the given
-     * {@code event} if its {@link TrackingToken} is covered by the previously scheduled event.
+     * Schedule a {@link MessageStream.Entry} for processing by this work package. Will immediately disregard the given
+     * {@code eventEntry} if its {@link TrackingToken} is covered by the previously scheduled event.
      * <p>
      * <b>Threading note:</b> This method is and should only to be called by the {@link Coordinator} thread of a {@link
      * PooledStreamingEventProcessor}.
      *
-     * @param event The event to schedule for work in this work package.
+     * @param eventEntry The event entry to schedule for work in this work package.
      * @return {@code True} if this {@link WorkPackage} scheduled the event for execution, otherwise {@code false}.
      */
-    public boolean scheduleEvent(TrackedEventMessage<?> event) {
-        if (shouldNotSchedule(event)) {
+    public boolean scheduleEvent(MessageStream.Entry<? extends EventMessage<?>> eventEntry) {
+        if (shouldNotSchedule(eventEntry)) {
+            TrackingToken eventToken = TrackingToken.fromContext(eventEntry).orElse(null);
             logger.trace("Ignoring event [{}] with position [{}] for work package [{}]. "
                                  + "The last token [{}] covers event's token [{}].",
-                         event.getIdentifier(), event.trackingToken().position().orElse(-1), segment.getSegmentId(),
-                         lastDeliveredToken, event.trackingToken());
+                         eventEntry.message().getIdentifier(),
+                         eventToken != null ? eventToken.position().orElse(-1) : -1,
+                         segment.getSegmentId(),
+                         lastDeliveredToken,
+                         eventToken);
             return false;
         }
-        logger.debug("Assigned event [{}] with position [{}] to work package [{}].",
-                     event.getIdentifier(), event.trackingToken().position().orElse(-1), segment.getSegmentId());
 
-        ProcessingContext context = new LegacyMessageSupportingContext(event);
-        boolean canHandle = canHandle(event, context);
-        processingQueue.add(new DefaultProcessingEntry(event, canHandle, context));
-        lastDeliveredToken = event.trackingToken();
+        TrackingToken eventToken = TrackingToken.fromContext(eventEntry).orElse(null);
```

Comment:
Nit: I think you can extract the `TrackingToken` once and reuse it in two spots. Right before the  `if (shouldNotSchedule(eventEntry)) {` invocation.

Reply:
You're right :) Refactored.